runtime/commonMain/src/kotlinx/serialization/SerialDescriptor.kt

```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * The number of elements this descriptor describes, besides from the element itself.
+     * [elementsCount] describes the number of **semantic** elements, not the number
+     * of actual fields/properties in the serialized form, even though they frequently match.
+     *
+     * For example, for the following class
+     * `class Complex(val real: Long, val imaginary: Long)` the corresponding descriptor
+     * and the serialized form both have two elements, while for `class IntList : ArrayList<Int>()`
+     * the corresponding descriptor has a single element (`IntDescriptor`, the type of list element),
+     * but from zero up to `Int.MAX_VALUE` elements in the serialized form.
+     */
+    public val elementsCount: Int
+
+    /**
+     * Returns a serial annotations of the associated class.
+     * Serial annotations can be used to specify an additional
+     * metadata that may be used during serialization, for example a [serial name][SerialName].
+     *
+     * TODO `serialAnnotations` or get rid of it completely?
+     * TODO default shouldn't be here either?
+     */
+    public val annotations: List<Annotation> get() = emptyList()
+
+    /**
+     * Returns a _positional_ name of the child at the given [index].
+     * Positional name usually represents a corresponding property name in the class, associated with
+     * the current descriptor.
+     *
+     * @throws IndexOutOfBoundsException for an illegal [index] values.
+     * @throws IllegalStateException if the current descriptor does not support children elements (e.g. is a primitive)
+     */
+    public fun getElementName(index: Int): String
+
+    /**
+     * Returns an index in the children list of the given element by its name or [CompositeDecoder.UNKNOWN_NAME]
+     * if there is no such element.
+     * The resulting index, if it is not [CompositeDecoder.UNKNOWN_NAME], is guaranteed to be usable with
+     * [getElementAnnotations], [getElementName] and [isElementOptional]
+     *
+     * TODO MapLike does not comply this contract
+     */
+    public fun getElementIndex(name: String): Int
+
+    /**
+     * Returns serial annotations of the child element at the given [index].
+     * This method differs from `getElementDescriptor(index).annotations` as it reports only
+     * declaration-specific annotations:
+     * ```
+     * @Serializable
+     * @SerialInfo("_nested")
+     * class Nested(...)
+     * @Serializable
+     * class Outer(@Polymorphic val nested: Nested)
+     *
+     * outerDescriptor.getElementAnnotations(0) // Returns [@Polymorphic]
+     * outerDescriptor.getElementDescriptor(0).annotations // Returns [@SerialInfo]
+     * ```
+     *
+     * @throws IndexOutOfBoundsException for an illegal [index] values.
+     * @throws IllegalStateException if the current descriptor does not support children elements (e.g. is a primitive)
+     */
+    public fun getElementAnnotations(index: Int): List<Annotation>
```

Comment:
why `annotations` have default value and this method does not have?

Reply:
Because `annotations` is unconditional property of the descriptor. Having an empty list default is completely reasonable.
For `getElementAnnotations` it is not, for the same reason `getElementIndex` does not default into `UNKNOWN_NAME`